pkgs/mkDepsCache.nix

```
@@ -43,9 +43,22 @@ let
     { lib, url, rev, hash, ... }:
     {
       name = "${lib}/${rev}";
-      path = fetchgit {
-        inherit url rev hash;
-      };
+      # we're using builtins.fetchGit, instead of fetchgit from
+      # build-support, so that we have seamless integration with
+      # ssh-agent or other credential mechanisms.
+      path = runCommand "${lib}/${rev}" {
+        src = builtins.fetchGit {
+          inherit url rev;
+          allRefs = true;
+        };
+        # we're wrapping builtins.fetchGit with a fixed-output
+        # derivation, re-using the same hash, that the fetchgit would
+        # use, from the prefetcher
+        outputHashMode = "recursive";
+        outputHash = hash;
+      } ''
+        cp -R $src $out
+      '';
```

Quite clever, but do you think we could use `fetchTree` here?   If `narHash` is set, it should first look in the nix store, e.g.:

```suggestion
      # we're using builtins.fetchTree, instead of fetchgit from
      # build-support, so that we have seamless integration with
      # ssh-agent or other credential mechanisms.
      # See https://nix.dev/manual/nix/latest/language/builtins.html#builtins-fetchTree
      path = builtins.fetchTree {
        type = "git";
        allRefs = true;
        narHash = hash;
        inherit url rev;
```

I only tested it with public git repos, could you verify if it works with a private repository? If not, I'll merge the current version

---

Very interesting, I was only vaguely aware of `fetchTree`, thanks! Let's check ...

---

```console
Checking out: forgejo@git.bendlas.net:webnf/datomic.git at 2502ddae61d754a23a241e23b2c2d85c2b6697e4
Checking out: forgejo@git.bendlas.net:webnf/enlive.git at a105d777e9ad05a19a910961370316870ff1bd1f
Checking out: forgejo@git.bendlas.net:webnf/filestore.git at bc0d7975c443739037771da92e465764a9bd0d28
```

It's doing the thing, is all I can say :-)

---

Yep, seems to be doing it, though.